8353290: C2: Refactor PhaseIdealLoop::is_counted_loop()

tabjy opened the pull request:
This PR refactors `PhaseIdealLoop::is_counted_loop()` into (mostly) `CountedLoopConverter::is_counted_loop()` and `CountedLoopConverter::convert()` to decouple to detection and conversion code. This enables us to try different loop configurations easy and finally convert once a counted loop is found. 

Blocks [JDK-8336759](https://bugs.openjdk.org/browse/JDK-8336759).

tabjy commented:
/reviewers 2

C2 changes

chhagedorn commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -310,7 +310,7 @@ IdealLoopTree* PhaseIdealLoop::insert_outer_loop(IdealLoopTree* loop, LoopNode*
 // clone the inner loop etc. No optimizations need to change the outer
 // strip mined loop as it is only a skeleton.
-IdealLoopTree* PhaseIdealLoop::create_outer_strip_mined_loop(BoolNode *test, Node *cmp, Node *init_control,
+IdealLoopTree* PhaseIdealLoop::create_outer_strip_mined_loop(Node *init_control,
```
Generally, for the touched code, you can also fix the `*` placement to be at the type.
```suggestion
IdealLoopTree* PhaseIdealLoop::create_outer_strip_mined_loop(Node* init_control,
```

tabjy replied:
Good point. I tried to fix them as I go but missed this one.

chhagedorn replied:
With the `x` -> `head` change, there are some more opportunities :-)

chhagedorn commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -364,7 +364,7 @@ void PhaseIdealLoop::insert_loop_limit_check_predicate(ParsePredicateSuccessProj
 }
 
-Node* PhaseIdealLoop::loop_exit_control(Node* x, IdealLoopTree* loop) {
+Node* PhaseIdealLoop::loop_exit_control(const Node* x, const IdealLoopTree* loop) const {
```
There are quite some places where use `x` to denote a `loop_head`. Maybe you can also fix that with this patch.

tabjy replied:
I would prefer a simple `head` for function named `.*loop.*`. Updated all relevant functions I touched.

chhagedorn commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -1468,47 +1462,49 @@ void PhaseIdealLoop::check_counted_loop_shape(IdealLoopTree* loop, Node* x, Basi
-  Node* phi_incr = nullptr;
-  incr = loop_iv_incr(incr, x, loop, phi_incr);
-  assert(incr != nullptr && incr->Opcode() == Op_Add(bt), "no incr");
+  LoopExitTest exit_test = loop_exit_test(back_control, loop);
```
Just an idea: Would it make sense to have a single `LoopStructure` (or another name) class that contains all the information stored now with separate structs `LoopExitTest`, `LoopIvIncr` and `LoopIvStride`? Then we could offer query methods for validation like `is_valid()`, `is_stride_valid()` etc., or check for existence like `has_stride()`, or access specific nodes like `stride()`, `incr()` etc.

tabjy replied:
Let me see what I can do! :)

chhagedorn commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -1266,26 +1270,92 @@ class PhaseIdealLoop : public PhaseTransform {
                                                IdealLoopTree* loop, float cl_prob, float le_fcnt,
                                                Node*& entry_control, Node*& iffalse);
 
+  class CountedLoopConverter {
```
Probably quite a subjective matter, but what about just naming it `CountedLoop`? Then you have `counted_loop.convert()`.

chhagedorn replied:
What I had in mind was something like that (I'm also fine with `CountedLoopConverter`:
```
CountedLoop counted_loop(...);
counted_loop.convert();
return counted_loop.is_valid();
```
But maybe you can explain in more detail what the follow-up work will be and how you use this class again later?

chhagedorn replied:
About the placement of this class. You could probably also move it out of the already huge `PhaseIdealLoop` class to make it a non-inner class, if there is not some strong coupling that you absolutely need.

tabjy replied:
> [...] more detail what the follow-up work [...]

I was referring to [JDK-8336759](https://bugs.openjdk.org/browse/JDK-8336759) and #22449. Essentially I'm trying to achieve:
```cpp
CountedLoopConverter c(...);
if (!c.is_counted_loop() && iv_bt == T_INT and && limit_t == T_LONG) {
    // transform `i < (long) limit` to `i < (int) (long) limit` if limit < INT_MAX
}

if (c.is_counted_loop()) {
    c.convert();
}
```

Which I personally think makes more sense to confirm a counted loops before trying out the conversion.

> [...] move it out of the already huge `PhaseIdealLoop` [...]

Yes, I agree! I found myself questioning this nested class, too.

chhagedorn commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -1266,26 +1270,92 @@ class PhaseIdealLoop : public PhaseTransform {
                                                Node*& entry_control, Node*& iffalse);
 
+  class CountedLoopConverter {
+    PhaseIdealLoop* _phase;
```
I suggest to add `const` whenever you can. For example, here you will probably not change the `_phase` pointer anymore:
```suggestion
    PhaseIdealLoop* const _phase;
```

tabjy replied:
That was my initial thought, too; however, `PhaseIdealLoop::insert_loop_limit_check_predicate()`, `::lazy_replace()`, `::set_subtree_ctrl()` and many other mutations makes this impossible.

chhagedorn replied:
You might be confusing it with having:
```
const PhaseIdealLoop* _phase
```
where we cannot call any non-const methods. That's indeed not possible. But what I mean was to make the pointer const:
```
PhaseIdealLoop* const _phase;
```
such that you cannot do `_phase = xyz` later. You would probably not do that anyway but it's an easy addition and safety for all fields not being reassigned again. It also helps to see which fields are going to be updated as part of the mutable state and which fields are not.

tabjy replied:
Sorry I neglected your diff. Yes `const` on the pointers make sense. I'm marking immutable fields `const`. Thanks for pointing this out!

chhagedorn commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -1266,26 +1270,92 @@ class PhaseIdealLoop : public PhaseTransform {
+    Node* _limit;
+    jlong _stride_con;
+    Node* _phi;
+    Node* _phi_incr;
+    Node* _stride;
+    bool _includes_limit;
+    BoolTest::mask _mask;
+    Node* _increment;
```
Here you have `(_phi_)incr` and `increment`. It should be consistent. I personally prefer full names whenever we can if it's not a universal abbreviation like `cmp` or `iv`. But what we count as well-known abbreviation is debatable and also depends on personal taste.

tabjy replied:
Updated `s/_phi_incr/_phi_increment`

chhagedorn commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -1266,26 +1270,92 @@ class PhaseIdealLoop : public PhaseTransform {
+    float _cl_prob;
+    Node* _sfpt;
+    jlong _final_correction;
+    Node* _trunc1;
```
`trunc1` is a little hard to understand. Can we find a better name?

tabjy replied:
Droped `Node* _trunc1` in favour of `TypeInteger* _increment_truncation_type`. Added a note in `should_stress_long_counted_loop()`.

chhagedorn reviewed (commented):
That's a good idea to refactor this, thanks for tackling it! Some initial general comments, I will have a closer look at your patch later again.

chhagedorn commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -1563,122 +1559,131 @@ bool PhaseIdealLoop::convert_to_long_loop(Node* cmp, Node* phi, IdealLoopTree* l
+  return false;
+}
+
+bool PhaseIdealLoop::CountedLoopConverter::is_counted_loop() {
```
`is_counted_loop()` is still huge. When trying to refactor it anyway, it might be a good idea to further split the code up into separate methods for all the different checks. That would increase the readability since we have a lot of different checks.

JohnTortugo commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -258,9 +258,13 @@ class CountedLoopNode : public BaseCountedLoopNode {
+    Node* trunc2 = nullptr;
+    const TypeInteger* trunc_type = nullptr;
+  };
+  static TruncatedIncrement match_incr_with_optional_truncation(Node* expr, BasicType bt);
```
Just a drive-by comment: this function and some others that you created are returning a value by copy, for performance reasons it may be better to return a reference or even a pointer as usually is the case in HotSpot.

tabjy replied:
I'm not sure if this is always the case but please correct me if I'm wrong: a cpp compiler already optimizes these kinds of patterns and structs are only allocated on the caller stack frame with a pointer to it passed into callee as an argument (or even leaving fields in registers if the struct is small enough). No copying is actually happening.

I agree this *syntactically looks costly*, please let me know if you find it necessary to switch to passing in a struct reference as an argument.

JohnTortugo commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -1266,26 +1270,91 @@ class PhaseIdealLoop : public PhaseTransform {
+  Node* loop_exit_control(const Node* head, const IdealLoopTree* loop) const;
+
+  struct LoopExitTest {
+    Node* cmp = nullptr;
```
I think it might be a good idea to use a more specific node type if you know before hand what kind of node the field will point to - in this case CmpNode*

JohnTortugo commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -1266,26 +1270,91 @@ class PhaseIdealLoop : public PhaseTransform {
+        _head(head),
+        _loop(loop),
+        _iv_bt(iv_bt) {
+      assert(phase != nullptr, "");
```
Would be nice to have an explanation of why the expression should be true.

chhagedorn commented:
Was out last week but I'm seeing your last commit mentions WIP. Let me know when it's ready to have another look again :-)

tabjy commented:
/touch

tabjy commented:
No worries! Enjoy your time off!

tabjy commented:
Resolved conflicts. @chhagedorn would kindly take a look? Thank you so much!

chhagedorn commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -2309,97 +2198,313 @@ bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_
+    return false;
+  }
+
+  PhaseIdealLoop::LoopExitTest exit_test = _phase->loop_exit_test(back_control, _loop);
```
Some thoughts/suggestions here:
- The method is still big and you need a moment to figure out what's going on/what checks we do.
- It looks like you are only initializing fields of `LoopStructure`. Couldn't you move the method to this class?
- You could have a separate field `_is_valid` in `LoopStructure`, then you could remove the `bool` return.. I.e. this would then look like this:
```
LoopStructure loop_structure;
loop_structure.build();
if (loop_structure.is_not_valid()) {
  return false;
}
```
You might need to pass in some additional info like `phase` to `LoopStructure` but I think that's okay.
- When doing the thing above, then you can just step by step assign the fields as you go. This would allow you to further split the method up which also improves documentation and moves field specific things to separate initializer methods:
```
back_control = _phase->loop_exit_control(_head, _loop);
if (back_control == nullptr) {
  _is_valid = false;
  return false;
}
exit_test = exit_test();
if (exit_test.is_not_valid()) {
  _is_valid = false;
  return;
}
incr = incr();
iv_incr = PhaseIdealLoop::loop_iv_incr(incr, _head, _loop);
....
```

chhagedorn commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -2309,97 +2198,313 @@ bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_
+
+  SafePointNode* sfpt = find_safepoint(back_control);
+
+  structure.back_control = back_control;
+
+  structure.cmp = exit_test.cmp;
+  structure.incr = incr;
+  structure.limit = exit_test.limit;
+  structure.mask = exit_test.mask;
+  structure.cl_prob = exit_test.cl_prob;
+
+  structure.iv_incr = iv_incr.incr;
+  structure.phi_incr = iv_incr.phi_incr;
+
+  structure.trunc_incr = increment.incr;
+  structure.trunc1 = increment.trunc1;
+  structure.trunc2 = increment.trunc2;
+  structure.trunc_type = increment.trunc_type;
+
+  structure.stride = stride.stride;
+
+  structure.phi = phi;
+
+  structure.sfpt = sfpt;
```
Are you later really going to use all these fields? I haven't double-checked. Another note here: I think it would be better to hide the fields and provide accessor methods. Otherwise, everyone can update them.

chhagedorn commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -1297,21 +1299,38 @@ class PhaseIdealLoop : public PhaseTransform {
+
+  static PhiNode* loop_iv_phi(const Node* xphi, const Node* phi_incr, const Node* head);
+
+  bool try_convert_to_counted_loop(Node* head, IdealLoopTree*&loop, const BasicType iv_bt);
```
```suggestion
  bool try_convert_to_counted_loop(Node* head, IdealLoopTree*& loop, const BasicType iv_bt);
```

chhagedorn commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -437,77 +439,71 @@ Node* PhaseIdealLoop::loop_exit_control(Node* x, IdealLoopTree* loop) {
 
-Node* PhaseIdealLoop::loop_exit_test(Node* back_control, IdealLoopTree* loop, Node*& incr, Node*& limit, BoolTest::mask& bt, float& cl_prob) {
-  Node* iftrue = back_control;
+PhaseIdealLoop::LoopExitTest PhaseIdealLoop::loop_exit_test(const Node* back_control, const IdealLoopTree* loop) {
```
Just an idea here: Could this also be part of `LoopExitTest` instead? Then a user could do something like:
```
LoopExitTest loop_exit_test(...);
 // i.e. = PhaseIdealLoop::loop_exit_test() but with a `_is_valid` flag. Then at the end you 
 // can also check for the right Cmp opcode and that it's not null which the current caller of 
 // loop_exit_test() are all doing. If that's off, you set `_is_valid` to false accordingly.
loop_exit_test.build();
if (loop_exit_test.is_not_valid()) {
...
}

```

chhagedorn reviewed (commented):
Hi @tabjy, sorry for letting you wait! I was very busy with other things. 

Thanks for coming back with an improved version! This looks much better already. I had a first over-viewing look. Will dive into it more again next week. I've just left some thoughts/comments here and there. Generally, I think we could improve on the classes to not just make them pure data holders with public access but actually allow users to call methods to interact with the data that we could hide to prevent modification. Let me know what you think :-)

tabjy commented:
@chhagedorn Thanks for reviewing. Sorry this took longer than I'd like. I've made the following changes:

- fixed typos
- better naming and pattern (e.g., `_` prefixes, unneeded code/brackets/spaces)
- changed `strcut` to `class` 
- hide fields behind accessor methods
- extracted relevant code into classes methods where makes sense
- resolved conflict

Thank you!

rwestrel commented on src/hotspot/share/opto/loopopts.cpp:
```diff
@@ -1696,9 +1692,7 @@ void PhaseIdealLoop::try_sink_out_of_loop(Node* n) {
-      !is_raw_to_oop_cast && // don't extend live ranges of raw oops
-      (KillPathsReachableByDeadTypeNode || !n->is_Type())
-      ) {
+      !n->is_Type()) {
```
This change seems unrelated. Bad merge?

tabjy commented:
@chhagedorn Sorry this took longer than expected. I left a few replies under some of your specific comments. All other issues were addressed. Thank you!

tabjy commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -1815,234 +1957,126 @@ bool PhaseIdealLoop::convert_to_long_loop(Node* cmp, Node* phi, IdealLoopTree* l
+  return false;
+}
 
-  // Trip-counter increment must be commutative & associative.
```
Done. Thank you for suggesting this.

chhagedorn commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -1720,47 +1638,279 @@ void PhaseIdealLoop::check_counted_loop_shape(IdealLoopTree* loop, Node* x, Basi
+void CountedLoopConverter::LoopStructure::build() {
+  _is_valid = false;
+
+  _back_control = _phase->loop_exit_control(_head, _loop);
```
You already call `loop_exit_control()` in the constructor. How about doing the following in the constructor instead?
```
_back_control(_phase->loop_exit_control(_head, _loop))
_exit_test(_back_control, _loop, _phase)
```

chhagedorn commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -1720,47 +1638,279 @@ void PhaseIdealLoop::check_counted_loop_shape(IdealLoopTree* loop, Node* x, Basi
+  }
+
+  // Should not reach
+  return false;
```
With the assert above for `stride_con` being 1 or -1, do we need this `if` here? Or are you concerned about when this does not hold in product? But since we are not using the return value anyways, I don't think it makes a difference and you could just drop the `if`.

chhagedorn commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -2584,21 +2729,24 @@ bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_
 // If there is one, then we do not need to create an additional Loop Limit Check Predicate.
-bool PhaseIdealLoop::has_dominating_loop_limit_check(Node* init_trip, Node* limit, const jlong stride_con,
+bool CountedLoopConverter::has_dominating_loop_limit_check(Node* init_trip, Node* limit, const jlong stride_con,
                                                      const BasicType iv_bt, Node* loop_entry) {
```
Can be made `const` and indentation is off.

tabjy commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -1720,47 +1638,279 @@ void PhaseIdealLoop::check_counted_loop_shape(IdealLoopTree* loop, Node* x, Basi
+  assert(incr->Opcode() == Op_AddI || incr->Opcode() == Op_AddL, "caller resp.");
+  // Get merge point
+  _xphi = incr->in(1);
+  _node = incr->in(2);
```
Renamed to `_stride_node` to avoid confusion between `jlong stride_con()`

tabjy commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -1720,47 +1638,279 @@ void PhaseIdealLoop::check_counted_loop_shape(IdealLoopTree* loop, Node* x, Basi
   Node* back_control = loop_exit_control(x, loop);
```
Not much, however, `is_valid()` is checking more conditions that are beyond the scope of a proper counted loop shape. I think it's a better idea to keep these `assert` as-is for easier debugging.

chhagedorn commented:
Thanks for the (ongoing?) updates! Let me know, when it's ready to be reviewed again :+1:

chhagedorn commented:
Was too busy this week, will try to come back to this next week!

chhagedorn commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -2231,78 +2249,62 @@ bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_
-  // Accounting for (LE2) and (LE4) where we use <= or >= in the loop exit check.
-  const bool includes_limit = (bt == BoolTest::le || bt == BoolTest::ge);
-  const jlong limit_correction_for_le_ge_exit_check = (includes_limit ? (stride_con > 0 ? 1 : -1) : 0);
+  int sov = check_stride_overflow(_structure.final_limit_correction(), limit_t, _iv_bt);
```
I suggest to rename it to `stride_overflow_state` or something like that since `sov` is a rather non-intuitive abbreviation. 

The best thing is probably to turn this into a proper enum since the states -1, 0, and 1 are not that easy to comprehend. I leave it up to you if you also want to do this in this PR - minor detail.

chhagedorn commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -2412,128 +2530,144 @@ bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_
+      _loop->_child == nullptr &&
+      _structure.sfpt() != nullptr &&
+      !_loop->_has_call &&
+      _phase->is_deleteable_safept(_structure.sfpt());
```
I think the old indentation was easier to read.

chhagedorn commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -2542,40 +2676,42 @@ bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_
 #endif
 
-  C->print_method(PHASE_AFTER_CLOOPS, 3, l);
+  _phase->C->print_method(PHASE_AFTER_CLOOPS, 3, l);
```
I was thinking about moving this below setting the phi type on the next lines to have this information already available in the IGV dump. I guess you could squeeze that change in here as well.

chhagedorn reviewed (commented):
Thanks for the updates! I first did some skimming first but ended up doing another complete pass. It's already in a good state but I still found some more small things which I left as comments. But then I think it's good to go from my side (regarding the code) :-)

About testing: 

Thanks for doing an extended testing with directly inserting the old code again to have a proper comparison. I first thought it's going to be too tricky which is why I proposed a logging - but I already feared that it's not going to be stable enough. So, I'm glad that you managed to do a old vs. new state!

For next steps, I suggest I'll give your patch a spin in our standard testing once you addressed the last comments in this badge. Then I'm also happy to run some more extended testing with your old vs. new counted loop transformation state (would be nice if you can update your branch with the latest review and also merge in latest master).

Let me know if you need some help :-)

tabjy commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -1720,47 +1638,269 @@ void PhaseIdealLoop::check_counted_loop_shape(IdealLoopTree* loop, Node* x, Basi
+  // until no more CastII/CastLL nodes are found.
+  while (xphi->Opcode() == Op_Cast(_iv_bt)) {
+    xphi = xphi->in(1);
+  }
```
Iterative uncasting was introduced in #25539, and I think that pr missed updating what's now `PhaseIdealLoop::check_counted_loop_shape()`, the only other caller of `xphi()`. `check_counted_loop_shape()` asserts after a loop nest is created. However,  [new test cases in #25539](https://github.com/openjdk/jdk/blob/8a0672c819e09a16c30fbdf58dc2b81f50958da4/test/hotspot/jtreg/compiler/loopopts/TestCountedLoopCastIV.java) does not create loop nests (if I understand correctly), and they didn't trigger this assertion which would otherwise fail.

It's safe to add this to `xphi()`. I'm not sure if created loop nests would contain this multiple-cast pattern at all, but it shouldn't break anything even they don't.

tabjy commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -1815,234 +1955,125 @@ bool PhaseIdealLoop::convert_to_long_loop(Node* cmp, Node* phi, IdealLoopTree* l
 
-  if (stride == nullptr) {
+  // Check trip counter will end up higher than the limit
+  const TypeInteger* limit_t = igvn->type(_structure.limit())->is_integer(_iv_bt);
```
moved to `is_infinite_loop()`

chhagedorn commented:
With your latest safepoint fix, the closed test is no longer failing! So, that seems to have fixed the issue. I'm currently running the DIFF patch with your fix on top up to tier7. Looking good so far. I'm also running some testing again for this patch only up to tier7. Will report back once it's complete (probably takes a while).

tabjy commented:
Hello @chhagedorn. Thank you so much for finding the root cause. I've updated the code, as you instructed, to generate necessary limit checks when stressing long loops. The only difference it that I do that in `CountedLoopConverter::stress_long_counted_loop()` instead of `CountedLoopConverter::convert()`.

I also included a new test file ([`TestStressLongCountedLoopLimitChecks.java`](https://github.com/openjdk/jdk/pull/24458/changes#diff-2a9525240bc695a8711f2905403d598213106aa3cd1d0647a449ce3e735928b7)) that tests against:
1. your reproducer and make sure init trip loop checks are inserted with `-XX:+TraceLoopLimitCheck`.
2. another case I constructed that needs both init trip and stride overflow limit checks to produce correct result

The master was merged in and conflicts were resolved. Thank you!

chhagedorn reviewed (commented):
Thanks for the update! A few minor comments. Afterwards, I can run some testing again. Can you also update the diff branch again? Then I can test that again as well.

krk commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -2340,97 +2382,205 @@ bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*& loop, BasicType iv
+  return false;
+}
+
+bool CountedLoopConverter::has_truncation_wrap(TruncatedIncrement truncation, Node* phi,
```
Drive-by comment. `has_truncation_wrap` could take a const reference: `const TruncatedIncrement&`, unless there is a specific reason for passing by value.

tabjy commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -1979,6 +2038,132 @@ class PhaseIdealLoop : public PhaseTransform {
+    PhaseIdealLoop::LoopIVStride& stride() { return _stride; }
+    PhiNode* phi() const { return _phi; }
+    SafePointNode* sfpt() const { return _safepoint; }
+    jlong stride_con() const { return _stride.compute_non_zero_stride_con(_exit_test.mask(), _iv_bt); }
```
In a non-debug build, `LoopStructure::stride_con()` was called once during counted loop detection and 1~3 times during conversion. Since I wanted to avoid caching (as counted loop strucutres can be mutable, as they will be in [JDK-8336759](https://github.com/openjdk/jdk/pull/22449)), I've updated `insert_*_limit_check()` taking an already resolved `stride_con`.

Now, we have **one** call during detection and **one** call during the actual conversion. Given the logic is relatively simple (a few logical comparisions), I don't belive this is a problem in terms of performance. 

Let me know what you think. Thanks for reviewing!

tabjy commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -1748,47 +1690,271 @@ void PhaseIdealLoop::check_counted_loop_shape(IdealLoopTree* loop, Node* x, Basi
+  }
+  assert(_truncated_increment.incr()->Opcode() == Op_Add(_iv_bt), "wrong increment code");
+
+  _stride.build(_truncated_increment.incr());
```
Similarly, some usage of `stride.build()` [uses a node from `iv_incr.incr()`](https://github.com/tabjy/jdk/blob/734afcbc78ec0c2eb3603bdb0cf4c67abbd57e02/src/hotspot/share/opto/loopnode.cpp#L1702) instead of `increment.incr()`.

rwestrel commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -1979,6 +2051,146 @@ class PhaseIdealLoop : public PhaseTransform {
+    const TypeInteger* trunc_type() const { return _trunc_type; }
+  };
+
+  class LoopStructure {
```
`public StackOb`j as well?

merykitty commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -1345,21 +1338,109 @@ class PhaseIdealLoop : public PhaseTransform {
-  PhiNode* loop_iv_phi(Node* xphi, Node* phi_incr, Node* x);
+  Node* loop_exit_control(const IdealLoopTree* loop) const;
+
+  class LoopExitTest {
```
I believe I just saw a JBS issue that says `StackObj` is becoming unnecessary and can be removed, so this can be optional.

https://bugs.openjdk.org/browse/JDK-8374333
